Feature/368 releases map

hoshmn opened the pull request:
hey @Lane - this is ready for you to have a look!

hoshmn commented on src/common/hooks/use-release-counties.js:
```diff
@@ -0,0 +1,67 @@
+
+export default function useReleaseCounties(data) {
+  const { allJailReleases, allPrisonReleases } = data;
+  // const { allJailReleases, allPrisonReleases } = useStaticQuery(
+  //   graphql`
+  //     query {
+  //       allJailReleases {
+  //         nodes {
+  //           jurisdiction
+  //         }
+  //       }
+  //       allPrisonReleases {
+  //         nodes {
+  //           jurisdiction
+  //         }
+  //       }
+  //     }
+  //   `
+  // );
```
not sure why this wasn't working, was getting this error:

> The result of this StaticQuery could not be fetched. This is likely a bug in Gatsby and if refreshing the page does not fix it, please open an issue in https://github.com/gatsbyjs/gatsby/issues

Lane replied:
i think that the `useStaticQuery` was failing because you need to import `graphql` from gatsby (new gatsby v3 requirement).  all good though, we don't want to use a static query for this data, we want it to be dynamic so it only fetches releases for the state (like you have done in the states template)

hoshmn replied:
mmm that makes sense. interesting that the error language isn't more specific...

hoshmn commented on src/templates/states.js:
```diff
@@ -23,6 +23,8 @@ import SectionNavigation from "../components/SectionNavigation";
 import content from "../../content/lang/states.json";
```
TODO: 
- clean up this file
- get rid of <Visual /> component
- tweak CSS and consider the DOM hierarchy (eg to include each map in the same card as its corresponding table)

hoshmn commented on src/templates/states.js:
```diff
@@ -194,29 +226,68 @@ const StateTemplate = ({ pageContext, data }) => {
               </div>
             </Step>
-
-            {scrollSections.map((section = {}, index) => {
```
it no longer seemed like it made sense to render this page by iterating over the `SECTION_COMPONENTS`, so now we render each component explicitly

hoshmn commented on src/common/hooks/use-release-counties.js:
```diff
@@ -0,0 +1,62 @@
+    // clone
+    const activeCounties = JSON.parse(JSON.stringify(allCounties));
+
+    // Hacky way to access the sole property, of form "cb_2015_puerto_rico_county_20m"
```
pretty sure DC is the only exception, and it's just a single county (with a single release in the county of "DC jail", so I guess strictly by our logic the county shouldn't be shaded)

hoshmn commented on src/common/hooks/use-release-counties.js:
```diff
@@ -0,0 +1,62 @@
+    const keys = Object.keys(allCounties.objects);
+    const countiesKey = keys[0];
+    // const name = stateName.toLowerCase().replace(" ", "_")
+    // const countiesKey = keys.find((k) => k.includes(name) && k.includes("county"))
```
maybe a slightly less hacky way of identifying the property? but i assume we can count on the form of the name not changing anyways...

hoshmn commented on src/components/maps/styles.js:
```diff
@@ -44,6 +44,24 @@ export const regionShapeStyles = (theme) => ({
+  shape: countyShapeStyleObj,
+  selected: {
+    fill: "#B3DB9A",
+    fillOpacity: "70%",
```
improvised on hover/selected styles since they weren't shown in figma (i modeled them after ice region styles)

Lane commented:
@hoshmn this is looking good!  a few minor changes:

- [ ]  **refactor `useReleaseCounties` into a function**

given the data is already being passed to the `useReleaseCounties` hook, i don't think this function needs to be a hook.  the `useMemo` won't lead to much performance gain (if any) because it gets called on almost every render, the tasks done within it aren't time consuming, and the change in value doesn't trigger any expensive re-renders.  without the need to use `useStaticQuery`, `useMemo`, or pull anything from the data stores this should be a utility function outside of the component instead of a hook.

*Example: `getActiveCounties` function in `ReleasesMap`*
```jsx

/**
 * Get the county shapes that have a release in the current state 
 */
const getActiveCounties = (data, stateName) => {
  const {allPrisonReleases, allJailReleases} = data;
  const state_code = getStateCodeByName(stateName);
  const stateCode = state_code.toUpperCase();
  const counties = [...allPrisonReleases.edges, ...allJailReleases.edges].map(
    (edge) => edge.node.jurisdiction.toLowerCase()
  );
  const allCounties = require(`../../maps/StateMap/topojson/${stateCode}.json`);
  // clone
  const activeCounties = JSON.parse(JSON.stringify(allCounties));
  // Hacky way to access the sole property, of form "cb_2015_puerto_rico_county_20m"
  const keys = Object.keys(allCounties.objects);
  const countiesKey = keys[0];
  if (!countiesKey || !activeCounties.objects[countiesKey]) return [];
  activeCounties.objects[countiesKey].geometries = activeCounties.objects[
    countiesKey
  ].geometries.filter(
    (geo) =>
      geo.properties.NAME &&
      counties.includes(geo.properties.NAME.toLowerCase())
  );
  return activeCounties
}

const ReleasesMap = ({ data, classes, ...props }) => {

//...

const activeCounties = getActiveCounties(data, stateName);

//...

}
```

- [ ] **add a "clear" button to the global filter (sitewide)**

whenever text is entered in the search, add a "clear" button (use the close icon) that removes the table filter.  whenever 

- [ ] **use a fixed size for the map**

map should not resize when selecting counties, set the map height relative to the viewport height instead of the section height.

Lane reviewed (changes requested):
a couple more change request i noticed:

- [x] do not clear the selected county when scrolling

![image](http://g.recordit.co/ZWsRqdrYmw.gif)

- [x] replace the blue box focus state with something more on-brand

currently clicking the map shows big blue borders around the area clicked.  can you switch this to the hover state or something similar?

hoshmn commented:
hey @Lane - thanks for the feedback! i think it should all be addressed, with the exception of combining the state maps in the same Step (/StepWrapper) as its corresponding table... not sure why i was having trouble getting that settled right now. but should be fine to demo without that and i can figure it out tomorrow (or if it's clear to you the best approach feel free)

Lane commented:
@hoshmn one more thing to add that was noticed during demo:

- [ ] use count formatter for "number of releases" column (1,234 instead of 1234)

hoshmn commented:
> @hoshmn one more thing to add that was noticed during demo:
> 
>     * [x]  use count formatter for "number of releases" column (1,234 instead of 1234)
good call. added the formatter and also refactored DefaultTable's modal to use the column's Cell renderer function (rather than just displaying `props.value`) so that numbers, links etc will be properly formatted in the modal as well
